src/workers/telegram_bot.rs

```diff
@@ -0,0 +1,312 @@
+) -> oneshot::Sender<()> {
+    let (shutdown_tx, shutdown_rx) = oneshot::channel();
+
+    tokio::spawn(async move {
+        listen_ngrok(auth_token, domain, forward_to, shutdown_rx).await;
+    });
+
+    shutdown_tx
+}
+
+async fn listen_ngrok(
+    auth_token: impl Into<String>,
+    domain: impl Into<String>,
+    forward_to: impl ToSocketAddrs + Send + Debug,
+    shutdown_rx: oneshot::Receiver<()>,
+) {
+    let session = ngrok::Session::builder().authtoken(auth_token).connect().await.unwrap();
+    let mut tunnel = session.http_endpoint().compression().domain(domain).listen().await.unwrap();
+
+    info!("Ngrok trying to forward tcp");
+    let url = tunnel.url().to_string();
+    info!("Ngrok forwarding to: {:?}", forward_to);
+    let tunnel = tunnel.forward_tcp(forward_to);
+    info!("Ngrok tunnel established at: {}", url);
+
+    tokio::select! {
+        _ = tunnel => {
+            info!("Ngrok tunnel closed");
+        }
+        _ = shutdown_rx => {
+            info!("Received shutdown signal, closing ngrok tunnel");
+        }
+    }
+}
+
+#[derive(BotCommands, Clone)]
+#[command(rename_rule = "lowercase")]
+enum Command {
+    #[command(description = "List commands.")]
+    Help,
+    #[command(description = "Get a photo from the camera.")]
+    Photo,
+    #[command(description = "Get a URL with video stream.")]
+    GetVideo,
+    #[command(description = "Stop video stream.")]
+    StopVideo,
+}
+
+/// Bot initialization and command handling
+async fn bot(context: &Context<WorkerMessage, Variables>) {
+    let telegram_config = &context.variables().telegram_config;
+    let bot = Bot::new(&telegram_config.token);
+
+    for command in Command::bot_commands() {
+        info!("Registered command: {:?}", command);
+    }
+
+    bot.set_my_commands(Command::bot_commands())
+        .await
+        .expect("Failed to set bot commands");
+
+    let handler = Update::filter_message()
+        .branch(dptree::entry().filter_command::<Command>().endpoint(command_handler))
+        .branch(
+            dptree::filter(|msg: Message| !msg.text().map_or(false, |text| text.starts_with('/')))
+                .endpoint(invalid_command_handler),
+        );
+
+    bot.send_message(ChatId(telegram_config.admin_user_id), "Bot started.")
+        .send()
+        .await
+        .expect("Failed to send message to admin user");
+
+    Dispatcher::builder(bot, handler)
+        .dependencies(dptree::deps![Arc::new(context.clone())])
+        .enable_ctrlc_handler()
+        .build()
+        .dispatch()
+        .await;
+}
+
+async fn command_handler(
+    bot: Bot,
+    msg: Message,
+    cmd: Command,
+    context: Arc<Context<WorkerMessage, Variables>>,
+) -> ResponseResult<()> {
+    let hc = context
+        .hub()
+        .register("bot: command handler", event_matches!(WorkerMessage::Frame(_)))
+        .unwrap();
+    let allowed_user_ids = &context.variables().telegram_config.allowed_user_ids;
+    match cmd {
```

Comment:
Maybe have reason to separate each case in single mod/method/file to perform better readability?

Reply:
yea, added refactor this function in general to my TODO, thanks)